okhttp-tests/src/test/java/com/squareup/okhttp/internal/ConnectionSpecSelectorTest.java

```diff
@@ -0,0 +1,153 @@
+          new SSLHandshakeException("Certificate handshake exception");
+      trustIssueException.initCause(new CertificateException());
+      boolean retry = connectionSpecSelector.connectionFailed(trustIssueException);
+      assertFalse(retry);
+    } finally {
+      socket.close();
+    }
+  }
+
+  @Test
+  public void retryableSSLHandshakeException() throws Exception {
+    ConnectionSpecSelector connectionSpecSelector =
+        createConnectionSpecSelector(ConnectionSpec.MODERN_TLS, ConnectionSpec.COMPATIBLE_TLS);
+    SSLSocket socket = createSocketWithEnabledProtocols(TlsVersion.TLS_1_1, TlsVersion.TLS_1_0);
+    try {
+      connectionSpecSelector.configureSecureSocket(socket);
+
+      boolean retry = connectionSpecSelector.connectionFailed(RETRYABLE_EXCEPTION);
+      assertTrue(retry);
+    } finally {
+      socket.close();
+    }
+  }
+
+  @Test
+  public void someFallbacksSupported() throws Exception {
+    ConnectionSpec sslV3 =
+        new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
+            .tlsVersions(TlsVersion.SSL_3_0)
+            .build();
+
+    ConnectionSpec[] connectionSpecs = {
+        ConnectionSpec.MODERN_TLS, ConnectionSpec.COMPATIBLE_TLS, sslV3
+    };
+    ConnectionSpecSelector connectionSpecSelector = createConnectionSpecSelector(connectionSpecs);
+
+    TlsVersion[] enabledSocketTlsVersions = { TlsVersion.TLS_1_1, TlsVersion.TLS_1_0 };
+    SSLSocket socket = createSocketWithEnabledProtocols(enabledSocketTlsVersions);
+
+    // MODERN_TLS is used here.
+    try {
+      connectionSpecSelector.configureSecureSocket(socket);
+      assertEnabledProtocols(socket, TlsVersion.TLS_1_1, TlsVersion.TLS_1_0);
+
+      boolean retry = connectionSpecSelector.connectionFailed(RETRYABLE_EXCEPTION);
+      assertTrue(retry);
+    } finally {
+      socket.close();
+    }
+
+    // COMPATIBLE_TLS is used here.
+    socket = createSocketWithEnabledProtocols(enabledSocketTlsVersions);
+    try {
+      connectionSpecSelector.configureSecureSocket(socket);
+      assertEnabledProtocols(socket, TlsVersion.TLS_1_0);
+
+      boolean retry = connectionSpecSelector.connectionFailed(RETRYABLE_EXCEPTION);
+      assertFalse(retry);
+    } finally {
+      socket.close();
+    }
+
+    // sslV3 is not used because SSLv3 is not enabled on the socket.
+  }
+
+  private static ConnectionSpecSelector createConnectionSpecSelector(
+      ConnectionSpec... connectionSpecs) {
+    return new ConnectionSpecSelector(Arrays.asList(connectionSpecs));
+  }
+
+  private SSLSocket createSocketWithEnabledProtocols(TlsVersion... tlsVersions) throws IOException {
+    SSLSocket socket = (SSLSocket) sslContext.getSocketFactory().createSocket();
+    socket.setEnabledProtocols(TlsVersion.javaNames(tlsVersions));
+    return socket;
+  }
+
+  private static void assertEnabledProtocols(SSLSocket socket, TlsVersion... required) {
+    Set<String> actual = new HashSet<String>(Arrays.asList(socket.getEnabledProtocols()));
```

Comment:
Never use regular `HashedSet`. Always `LinkedHashSet`. The cost is negligible, and the predictability is invaluable.